File: Single-Linked-List/singly_linked_list_traverse_reverse.c
// File Name: singly_linked_list_traverse_reverse.c
// Program to traverse and display a singly linked list in reverse order

#include <stdio.h>
#include <stdlib.h>

// Definition of node
struct node {
    int data;
    struct node *next;
};

struct node *head = NULL;

// Recursive function to traverse in reverse
void traverse_reverse(struct node *temp) {
    if (temp == NULL)
        return;

    traverse_reverse(temp->next);
    printf("%d -> ", temp->data);
}

int main() {
    if (head == NULL) {
        printf("List is empty.\n");
    } else {
        printf("Reverse order: ");
        traverse_reverse(head);
        printf("NULL\n");
    }
    return 0;
}

File: Single-Linked-List/SinglyLinkedListTraverseReverse.java
// File Name: SinglyLinkedListTraverseReverse.java
// Program to traverse and display a singly linked list in reverse order

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class SinglyLinkedListTraverseReverse {
    static Node head = null;

    // Recursive function to traverse in reverse
    static void traverse_reverse(Node temp) {
        if (temp == null)
            return;

        traverse_reverse(temp.next);
        System.out.print(temp.data + " -> ");
    }

    public static void main(String[] args) {
        if (head == null) {
            System.out.println("List is empty.");
        } else {
            System.out.print("Reverse order: ");
            traverse_reverse(head);
            System.out.println("NULL");
        }
    }
}
